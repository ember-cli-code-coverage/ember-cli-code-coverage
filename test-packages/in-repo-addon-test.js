'use strict';

const fs = require('fs-extra');
const util = require('util');
const rimraf = util.promisify(require('rimraf'));
const chai = require('chai');
const expect = chai.expect;
const chaiFiles = require('chai-files');
const dir = chaiFiles.dir;
const file = chaiFiles.file;
const path = require('path');
const execa = require('execa');

chai.use(chaiFiles);

const BASE_PATH = path.join(__dirname, 'my-app-with-in-repo-addon');

describe('in-repo addon coverage generation', function () {
  this.timeout(10000000);

  beforeEach(async function () {
    await rimraf(`${BASE_PATH}/coverage*`);
    await execa('git', ['clean', '-f', 'my-app-with-in-repo-addon'], { cwd: __dirname });
    await execa('git', ['restore', 'my-app-with-in-repo-addon'], { cwd: __dirname });
  });

  afterEach(async function () {
    await rimraf(`${BASE_PATH}/coverage*`);
    await execa('git', ['clean', '-f', 'my-app-with-in-repo-addon'], { cwd: __dirname });
    await execa('git', ['restore', 'my-app-with-in-repo-addon'], { cwd: __dirname });
  });

  it('runs coverage on in-repo addon', async function () {
    expect(dir(`${BASE_PATH}/coverage`)).to.not.exist;

    let env = { COVERAGE: 'true' };
    await execa('ember', ['test'], { cwd: BASE_PATH, env });
    expect(file(`${BASE_PATH}/coverage/lcov-report/index.html`)).to.not.be.empty;
    expect(file(`${BASE_PATH}/coverage/index.html`)).to.not.be.empty;

    const summary = fs.readJSONSync(`${BASE_PATH}/coverage/coverage-summary.json`);
    expect(summary.total.lines.pct).to.equal(46.67);
    expect(summary['app/utils/my-covered-util-app.js'].lines.total).to.equal(1);

    // Check that lib/my-in-repo-addon/utils/my-covered-utill is 1 line and that 1 line is covered
    expect(summary['lib/my-in-repo-addon/addon/utils/my-covered-util.js'].lines.total).to.equal(1);
    expect(summary['lib/my-in-repo-addon/addon/utils/my-covered-util.js'].lines.covered).to.equal(
      1
    );

    // Check that lib/my-in-repo-addon/utils/my-uncovered-utill is 1 line and that 0 lines are covered
    expect(summary['lib/my-in-repo-addon/addon/utils/my-uncovered-util.js'].lines.total).to.equal(
      1
    );
    expect(summary['lib/my-in-repo-addon/addon/utils/my-uncovered-util.js'].lines.covered).to.equal(
      0
    );

    // Check that lib/my-in-repo-addon/addon-test-support/uncovered-test-support is 4 lines and that 0 lines are covered
    expect(
      summary['lib/my-in-repo-addon/addon-test-support/uncovered-test-support.js'].lines.total
    ).to.equal(4);
    expect(
      summary['lib/my-in-repo-addon/addon-test-support/uncovered-test-support.js'].lines.covered
    ).to.equal(0);
  });
});
