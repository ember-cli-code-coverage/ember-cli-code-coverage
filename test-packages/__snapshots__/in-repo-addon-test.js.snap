// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`in-repo addon coverage generation runs coverage on in-repo addon 1`] = `
Object {
  "app/app.js": Object {
    "branches": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "lines": Object {
      "covered": 4,
      "pct": 100,
      "skipped": 0,
      "total": 4,
    },
    "statements": Object {
      "covered": 4,
      "pct": 100,
      "skipped": 0,
      "total": 4,
    },
  },
  "app/router.js": Object {
    "branches": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "lines": Object {
      "covered": 1,
      "pct": 33.33,
      "skipped": 0,
      "total": 3,
    },
    "statements": Object {
      "covered": 1,
      "pct": 33.33,
      "skipped": 0,
      "total": 3,
    },
  },
  "app/services/my-service.js": Object {
    "branches": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
    "lines": Object {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
    "statements": Object {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
  },
  "app/utils/my-covered-util-app.js": Object {
    "branches": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
    "lines": Object {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
    "statements": Object {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
  },
  "app/utils/my-uncovered-util-app.js": Object {
    "branches": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "lines": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "statements": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
  },
  "lib/my-in-repo-addon/addon-test-support/uncovered-test-support.js": Object {
    "branches": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 2,
    },
    "functions": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 2,
    },
    "lines": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 4,
    },
    "statements": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 4,
    },
  },
  "lib/my-in-repo-addon/addon/services/my-service.js": Object {
    "branches": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "lines": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "statements": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
  },
  "lib/my-in-repo-addon/addon/utils/my-covered-util.js": Object {
    "branches": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
    "lines": Object {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
    "statements": Object {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
  },
  "lib/my-in-repo-addon/addon/utils/my-uncovered-util.js": Object {
    "branches": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "lines": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "statements": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
  },
  "lib/my-in-repo-addon/app/utils/my-covered-util.js": Object {
    "branches": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 0,
    },
    "lines": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 0,
    },
    "statements": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 0,
    },
  },
  "lib/my-in-repo-addon/app/utils/my-uncovered-util.js": Object {
    "branches": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 0,
    },
    "lines": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 0,
    },
    "statements": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 0,
    },
  },
  "total": Object {
    "branches": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 2,
    },
    "functions": Object {
      "covered": 3,
      "pct": 33.33,
      "skipped": 0,
      "total": 9,
    },
    "lines": Object {
      "covered": 8,
      "pct": 47.06,
      "skipped": 0,
      "total": 17,
    },
    "statements": Object {
      "covered": 8,
      "pct": 47.06,
      "skipped": 0,
      "total": 17,
    },
  },
}
`;
